hack/tools/release/testimage.sh

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function TESTIMAGE() {
+  IMAGE="${1}"
+  ARCH="${2}"
+  GREPGOPATH="${3}"
+  BINARYPATH="${4:-manager}"
+
+  echo "> Testing $IMAGE"
+
+  docker save "${IMAGE}" -o img.tar
+  mkdir -p extracted-img extracted
+  tar -xf img.tar -C extracted-img/
+  while IFS= read -r -d '' layer
+  do
+    tar -xf "${layer}" -C extracted
+  done <   <(find extracted-img/ -name "*.tar" -print0)
+
+  go version -m extracted/"${BINARYPATH}" | grep -P "\tpath" | grep -E -q -e "${GREPGOPATH}" || (echo "FAILED ${IMAGE} expected value for path: \"${GREPGOPATH}\""; go version -m extracted/"${BINARYPATH}" | grep -P "\tpath")
+  go version -m extracted/"${BINARYPATH}" | grep -q -E "GOARCH=${ARCH}$" || (echo "Failed ${IMAGE} expected GOARCH to be \"$ARCH\""; go version -m extracted/"${BINARYPATH}" | grep "GOARCH=")
+
+  rm -rf img.tar extracted-img extracted
+}
+
+for arch in ${ALL_ARCH}; do
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/cluster-api-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api$"
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/kubeadm-bootstrap-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/bootstrap/kubeadm$"
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/kubeadm-control-plane-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/controlplane/kubeadm$"
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/capd-manager-${arch}:dev" "${arch}" "command-line-arguments$"
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/capim-manager-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/test/infrastructure/inmemory$"
+  TESTIMAGE "gcr.io/k8s-staging-cluster-api/test-extension-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/test/extension$"
```

Comment:
We have to use `REGISTRY` here:

```suggestion
  TESTIMAGE "${REGISTRY}/cluster-api-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api$"
  TESTIMAGE "${REGISTRY}/kubeadm-bootstrap-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/bootstrap/kubeadm$"
  TESTIMAGE "${REGISTRY}/kubeadm-control-plane-controller-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/controlplane/kubeadm$"
  TESTIMAGE "${REGISTRY}/capd-manager-${arch}:dev" "${arch}" "command-line-arguments$"
  TESTIMAGE "${REGISTRY}/capim-manager-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/test/infrastructure/inmemory$"
  TESTIMAGE "${REGISTRY}/test-extension-${arch}:dev" "${arch}" "sigs.k8s.io/cluster-api/test/extension$"
```